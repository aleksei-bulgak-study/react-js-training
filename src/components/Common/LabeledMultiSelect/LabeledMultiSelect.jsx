import React from 'react';
import PropTypes from 'prop-types';
import MultiSelect from 'react-multi-select-component';
import './LabeledMultiSelect.css';

const buildValues = (elems) => elems.map((e) => ({ label: e, value: e }));

const LabeledMultiSelect = ({ title, options, onAction, preselected, error }) => {
  const selected = buildValues(preselected) || [];

  return (
    <label htmlFor="genres" className="genres">
      {title}
      <MultiSelect
        id="genres"
        options={options}
        value={selected}
        onChange={onAction}
        hasSelectAll={false}
        disableSearch
      />
      {error !== '' && <p className="genres__error">{error}</p>}
    </label>
  );
};

LabeledMultiSelect.propTypes = {
  title: PropTypes.string.isRequired,
  onAction: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    }),
  ).isRequired,
  preselected: PropTypes.arrayOf(PropTypes.string),
  error: PropTypes.string,
};

LabeledMultiSelect.defaultProps = {
  preselected: [],
  error: '',
};

export default LabeledMultiSelect;
