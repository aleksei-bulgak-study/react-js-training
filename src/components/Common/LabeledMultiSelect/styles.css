.genres .multi-select,
.genres .dropdown-container {
  --rmsc-bg: #2E2E2E;
  --rmsc-hover: #2E2E2E;
  --rmsc-selected: #2E2E2E;
  background-color: #2E2E2E;
  outline: none;
  color: var(--white);
  border: none;
  border-radius: none;
  color: #B7B7B7;
}

.genres .item-renderer {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
}

.genres .item-renderer input {
  width: 1rem;
}

.genres .item-renderer span {
  color: #B7B7B7;
  text-transform: capitalize;
}

.genres__error {
  font-size: 0.8rem;
  font-weight: 100;
  color: var(--red);
  text-transform: none;
}
