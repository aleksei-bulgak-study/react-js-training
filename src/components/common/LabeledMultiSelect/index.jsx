import React, { useState } from 'react';
import PropTypes from 'prop-types';
import MultiSelect from 'react-multi-select-component';
import './styles.css';

const buildValues = (elems) => elems.map((e) => ({ label: e, value: e }));

const LabeledMultiSelect = ({ title, options, onAction, preselected }) => {
  const [selected, setSelected] = useState(buildValues(preselected) || []);

  const onChange = (selectedElements) => {
    onAction(selectedElements);
    setSelected(selectedElements);
  };

  return (
    <label htmlFor="genres" className="genres">
      {title}
      <MultiSelect
        id="genres"
        options={options}
        value={selected}
        onChange={onChange}
        hasSelectAll={false}
        disableSearch
      />
    </label>
  );
};

LabeledMultiSelect.propTypes = {
  title: PropTypes.string.isRequired,
  onAction: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
    }),
  ).isRequired,
  preselected: PropTypes.arrayOf(PropTypes.string),
};

LabeledMultiSelect.defaultProps = {
  preselected: [],
};

export default LabeledMultiSelect;
