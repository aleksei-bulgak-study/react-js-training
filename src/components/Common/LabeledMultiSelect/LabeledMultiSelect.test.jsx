import React from 'react';
import { mount } from 'enzyme';
import LabeledMultiSelect from './LabeledMultiSelect';

describe('LabeledMultiSelect', () => {
  test('when error was specified then it rendered', () => {
    const onAction = jest.fn();
    const select = mount(
      <LabeledMultiSelect
        title="test title"
        onAction={onAction}
        name="testName"
        error="error message"
        options={[{ label: 'Test', value: 'test' }]}
      />,
    );
    expect(select.find('p.genres__error').text()).toEqual(
      'error message',
    );
    expect(select.html()).toMatchSnapshot();
  });
});
