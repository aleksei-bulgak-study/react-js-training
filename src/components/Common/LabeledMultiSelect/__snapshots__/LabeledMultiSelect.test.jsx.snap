// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabeledMultiSelect when error was specified then it rendered 1`] = `"<label for=\\"genres\\" class=\\"genres\\">test title<div class=\\"go2139093995 multi-select\\"><div tabindex=\\"0\\" class=\\"go2646822163 dropdown-container\\" aria-expanded=\\"false\\" aria-readonly=\\"true\\"><div class=\\"go2642161244 dropdown-heading\\"><div class=\\"dropdown-heading-value\\"><span class=\\"gray\\">Select...</span></div><svg width=\\"24\\" height=\\"24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" class=\\"dropdown-heading-dropdown-arrow gray\\"><path d=\\"M6 9L12 15 18 9\\"></path></svg></div></div></div><p class=\\"genres__error\\">error message</p></label>"`;
